### PR TITLE
Feature: SoftMax, CosFace, ArcFace layers

### PR DESCRIPTION
**Pull request by ditwoo**
## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [x] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [x] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [x] Did you check the docs with `make check-docs`?
- [x] Did you write any new necessary tests?
- [x] Did you add your new functionality to the docs?
- [x] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/CHANGELOG.md)?
- [x] **You can use 'Login as guest' to see Teamcity build logs.**

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: catalyst/contrib/nn/modules/__init__.py [hunk 1]
```diff
@@ -31,3 +31,7 @@
     scSE,
     cSE,
 )
+
+from catalyst.contrib.nn.modules.softmax import SoftMax
+from catalyst.contrib.nn.modules.arcface import ArcFace
+from catalyst.contrib.nn.modules.cosface import CosFace
```
Comment (Scitator) [inline catalyst/contrib/nn/modules/__init__.py]: as far as this is criterions, could we move them to `contrib/nn/criterion`? what do you think about this?

File: catalyst/contrib/nn/modules/arcface.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        one_hot = torch.zeros(cos_theta.size()).to(input.device)
+        one_hot.scatter_(1, label.view(-1, 1).long(), 1)
+
+        selected = torch.where(
+            theta > self.threshold, torch.zeros_like(one_hot), one_hot
+        )
+
+        output = torch.cos(torch.where(selected.bool(), theta + self.m, theta))
+        output *= self.s
+        return output
```
Comment (Scitator) [inline catalyst/contrib/nn/modules/arcface.py]: could you please rename `output` to `logits` for readability? In the docs you name it `logits`

File: catalyst/contrib/nn/modules/softmax.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class SoftMax(nn.Module):
+    """Implementation of SoftMax head for metric learning.
```
Comment (Scitator) [inline catalyst/contrib/nn/modules/softmax.py]: article link?)
